### PR TITLE
Add lookback window support on trigger registration

### PR DESCRIPTION
**Pull request by johnivdel**
Resolves https://github.com/WICG/attribution-reporting-api/issues/769

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/pull/833.html" title="Last updated on Jun 6, 2023, 9:12 PM UTC (6a1d5a9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/833/62fe978...6a1d5a9.html" title="Last updated on Jun 6, 2023, 9:12 PM UTC (6a1d5a9)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -2018,6 +2035,37 @@ ISSUE: properly define the "finishing issuance" operation.
...
+        [=parse an optional 64-bit unsigned integer=] with |value|,
+        "`window`", and null.
+    1. If |window| is an error, return null.
+    1. Let |filters| be a new [=filter map=].
+    1. If |values|["`filters`"] [=map/exists=]:
+        1. Set |filters| to the result of running [=parse filter data=] with |value|["`filters`"].
+        1. If |filters| is null, return null.
+    1. Let |negatedFilters| be a new [=filter map=].
+    1. If |values|["`not_filters`"] [=map/exists=]:
+        1. Set |negatedFilters| to the result of running [=parse filter data=] with |value|["`not_filters`"].
```
Comment (apasel422) [inline index.bs:2058]: These should use https://wicg.github.io/attribution-reporting-api/#parse-filters not "parse filter data".

File: index.bs [hunk 2]
```diff
@@ -2018,6 +2035,37 @@ ISSUE: properly define the "finishing issuance" operation.
...
+    
+To <dfn>parse lookback windows</dfn> given an [=ordered map=] |map|:
+
+1. Let |lookbackWindows| be a new [=list=].
+1. If |map|["`lookback_windows`"] does not [=map/exist=], return |aggregatableDedupKeys|.
+1. Let |values| be |map|["`lookback_windows`"].
+1. If |values| is not a [=list=], return null.
+1. [=list/iterate|For each=] |value| of |values|:
+    1. If |value| is not an [=ordered map=], return null.
+    1. If |value|["`window`"] does not [=map/exist=] or is not a [=string=], return null.
```
Comment (apasel422) [inline index.bs:2047]: I think this check is redundant, as we already provide `null` as the default to `parse an optional...` below, so we can just check "If |window| is null or an error, return null."

File: index.bs [hunk 3]
```diff
@@ -668,6 +668,20 @@ An aggregatable dedup key is a [=struct=] with the following items:
...
 
+</dl>
+        
+<h3 dfn-type=dfn>Lookback window</h3>
+
+A lookback window is a [=struct=] with the following items:
+
+<dl dfn-for="lookback window">
+: <dfn>window</dfn>
+:: a [=duration=].
```
Comment (apasel422) [inline index.bs:679]:
```suggestion
:: A [=duration=].
```

Comment (apasel422) [inline index.bs:679]: Should this be non-negative or positive?

File: index.bs [hunk 4]
```diff
@@ -2018,6 +2035,37 @@ ISSUE: properly define the "finishing issuance" operation.
...
+        1. Set |negatedFilters| to the result of running [=parse filter data=] with |value|["`not_filters`"].
+        1. If |negatedFilters| is null, return null.
+    1. Let  |lookbackWindow| be a new [=lookback window=] with the items:
+        : [=lookback window/window=]
+        :: |window| seconds
+        : [=lookback window/filters=]
+        :: |filters|
+        : [=lookback window/negated filters=]
+        :: |negatedFilters|
+    1. [=set/Append=] |lookbackWindow| to |lookbackWindows|.
```
Comment (apasel422) [inline index.bs:2067]: This is a set operation, but `lookbackWindows` is a list.

File: index.bs [hunk 5]
```diff
@@ -2073,6 +2121,7 @@ and a [=moment=] |triggerTime|:
 1. [=list/iterate|For each=] |privateStateToken| of |privateStateTokens|:
     1. Let |serializedPrivateStateToken| be the result of [=serializing a private state token=] with |privateStateToken|.
     1. [=list/Append=] |serializedPrivateStateToken| to  |serializedPrivateStateTokens|.
+1. Let |lookbackWindows| be the result of running [=parse lookback windows=] with |value|.
```
Comment (apasel422) [inline index.bs:2124]: We need to check if this is null and return early if so.

File: index.bs [hunk 6]
```diff
@@ -2162,6 +2213,16 @@ To <dfn>match an attribution source's filter data against filters and negated fi
     |source|, |notFilters|, and [=match an attribution source's filter data against filters/isNegated=] set to true is false, return false.
 1. Return true.
 
+<h3 dfn id="check-source-in-window">Check if source is in lookback window</h3>
+    
+To  <dfn>check if source is in lookback window</dfn> given an [=attribution source=] |source| and [=attribution trigger=] |trigger|:
```
Comment (apasel422) [inline index.bs:2218]:
```suggestion
To  <dfn>check if source is in lookback window</dfn> given an [=attribution source=] |source| and an [=attribution trigger=] |trigger|:
```

File: index.bs [hunk 7]
```diff
@@ -2554,6 +2615,15 @@ To <dfn noexport>trigger attribution</dfn> given an [=attribution trigger=] |tri
     1. If |hasAggregatableData| is true, then run [=generate null reports and assign private state tokens=]
         with |trigger| and [=generate null reports and assign private state tokens/report=] set to null.
     1. Return.
+1. If the result of running
+    [=check if source is in lookback window=] with
+    |sourceToAttribute| and |trigger| is false,
```
Comment (apasel422) [inline index.bs:2620]:
```suggestion
    |sourceToAttribute| and |trigger| is false:
```

**Review by apasel422 — CHANGES_REQUESTED** [review]

**Comment by apasel422** [comment]
Needs rebase.

**Comment by csharrison** [comment]
This is obsolete with #914 right? Can we close it?

**Comment by apasel422** [comment]
Closing in favor of #914

**Pull request closed without merging**